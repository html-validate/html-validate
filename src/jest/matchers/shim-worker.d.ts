declare module "*?worker&url" {
	const value: string;
	export default value;
}
