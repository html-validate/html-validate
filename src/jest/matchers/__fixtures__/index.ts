export { reportOk, reportOkAsync } from "./report-ok";
export { reportError, reportErrorAsync } from "./report-error";
export { reportMultipleErrors, reportMultipleErrorsAsync } from "./report-multiple-errors";
