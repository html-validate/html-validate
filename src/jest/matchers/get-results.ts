import { FileSystemConfigLoader } from "../../config/loaders/file-system";
import { type Report, type Result } from "../../reporter";
import HtmlValidate from "../../htmlvalidate";

/**
 * @internal
 */
export function getResults(filename: string, value: Report | string): Result[] {
	if (typeof value === "string") {
		const loader = new FileSystemConfigLoader({
			extends: ["html-validate:recommended"],
		});
		const htmlvalidate = new HtmlValidate(loader);
		const report = htmlvalidate.validateString(value, filename, {
			rules: {
				"void-style": "off",
			},
		});
		return report.results.map((it) => {
			return { ...it, filePath: "inline" };
		});
	} else {
		return value.results;
	}
}
