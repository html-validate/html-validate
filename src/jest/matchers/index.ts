export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
