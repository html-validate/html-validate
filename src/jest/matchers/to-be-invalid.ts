import { Report } from "../../reporter";
import { diverge } from "../utils";

function toBeInvalid(report: Report): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: false,
			message: () => "Result should be invalid but had no errors",
		};
	} else {
		return {
			pass: true,
			message: /* istanbul ignore next */ () => "Result should not contain error",
		};
	}
}

export default diverge(toBeInvalid);
