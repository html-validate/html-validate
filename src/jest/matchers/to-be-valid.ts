import { Report } from "../../reporter";

export async function toBeValid(
	this: jest.MatcherUtils,
	actual: Report | Promise<Report>
): Promise<jest.CustomMatcherResult> {
	const report = await actual;
	if (report.valid) {
		return {
			pass: true,
			message: /* istanbul ignore next */ () => "Result should not contain error",
		};
	} else {
		const firstError = report.results[0].messages[0];
		return {
			pass: false,
			message: () => `Result should be valid but had error "${firstError.message}"`,
		};
	}
}
