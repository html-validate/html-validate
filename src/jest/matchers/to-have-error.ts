import { type Message, type Report } from "../../reporter";
import { type MatcherContext, type MatcherResult, diff, diverge } from "../utils";
import { flattenMessages } from "../utils/flatten-messages";

function toHaveErrorImpl(
	jest: MatcherContext,
	actual: Report,
	expected: Partial<Message>
): MatcherResult {
	const flattened = flattenMessages(actual);
	const matcher = [expect.objectContaining(expected)];
	const pass = jest.equals(flattened, matcher);
	const diffString = diff(matcher, flattened, { expand: jest.expand });
	const hint = jest.utils.matcherHint(".toHaveError");
	const prettyExpected = jest.utils.printExpected(matcher);
	const prettyReceived = jest.utils.printReceived(flattened);
	const resultMessage = (): string => {
		return [
			hint,
			"",
			"Expected error to equal:",
			`  ${prettyExpected}`,
			"Received:",
			`  ${prettyReceived}`,
			/* istanbul ignore next */ diffString ? `\nDifference:\n\n${diffString}` : "",
		].join("\n");
	};
	return { pass, message: resultMessage };
}

function toHaveError(this: MatcherContext, actual: Report, error: Partial<Message>): MatcherResult;
function toHaveError(
	this: MatcherContext,
	actual: Report,
	ruleId: string,
	message: string,
	context?: any
): MatcherResult;
function toHaveError(
	this: MatcherContext,
	actual: Report,
	arg1: string | Partial<Message>,
	arg2?: string,
	arg3?: any
): MatcherResult {
	if (typeof arg1 === "string") {
		const expected: Partial<Message> = {
			ruleId: arg1,
			message: arg2,
		};
		if (arg3) {
			/* eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- this is supposed to accept anything */
			expected.context = arg3;
		}
		return toHaveErrorImpl(this, actual, expected);
	} else {
		return toHaveErrorImpl(this, actual, arg1);
	}
}

export default diverge(toHaveError);
