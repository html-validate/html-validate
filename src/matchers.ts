/* eslint-disable @typescript-eslint/no-namespace, @typescript-eslint/ban-ts-comment, prefer-template, sonarjs/no-duplicate-string */

import diff from "jest-diff";
import deepmerge from "deepmerge";
import { TokenType } from "./lexer";
import { Message, Report, Result } from "./reporter";
import { ConfigData } from "./config";
import HtmlValidate from "./htmlvalidate";

interface TokenMatcher {
	type: TokenType;
	location?: any;
	data?: any;
}

declare global {
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toBeToken(expected: TokenMatcher): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(
				errors: Array<[string, string] | Record<string, unknown>>
			): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 */
			toHTMLValidate(config?: ConfigData): R;
		}
	}
}

/**
 * Takes all messages from all files and flattens to a single array.
 */
function flattenMessages(report: Report): Message[] {
	return report.results.reduce((aggregated: Message[], result: Result) => {
		return aggregated.concat(result.messages);
	}, []);
}

function toBeValid(
	this: jest.MatcherUtils,
	report: Report
): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: true,
			message: /* istanbul ignore next */ () =>
				"Result should not contain error",
		};
	} else {
		const firstError = report.results[0].messages[0];
		return {
			pass: false,
			message: () =>
				`Result should be successful but had error "${firstError.message}"`,
		};
	}
}

function toBeInvalid(
	this: jest.MatcherUtils,
	report: Report
): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: false,
			message: () => "Result should be invalid but had no errors",
		};
	} else {
		return {
			pass: true,
			message: /* istanbul ignore next */ () =>
				"Result should not contain error",
		};
	}
}

function toHaveError(
	this: jest.MatcherUtils,
	report: Report,
	ruleId: any,
	message: any,
	context?: any
): jest.CustomMatcherResult {
	const actual = flattenMessages(report);
	const expected: any = { ruleId, message };
	if (context) {
		expected.context = context;
	}

	const matcher = [expect.objectContaining(expected)];
	const pass = this.equals(actual, matcher);
	const diffString = diff(matcher, actual, { expand: this.expand });
	const resultMessage = (): string =>
		this.utils.matcherHint(".toHaveError") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(actual)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message: resultMessage };
}

function toHaveErrors(
	this: jest.MatcherUtils,
	report: Report,
	errors: Array<[string, string] | Record<string, unknown>>
): jest.CustomMatcherResult {
	const actual = flattenMessages(report);
	const matcher = errors.map((entry) => {
		if (Array.isArray(entry)) {
			const [ruleId, message] = entry;
			return expect.objectContaining({ ruleId, message });
		} else {
			return expect.objectContaining(entry);
		}
	});
	const pass = this.equals(actual, matcher);
	const diffString = diff(matcher, actual, { expand: this.expand });
	const resultMessage = (): string =>
		this.utils.matcherHint(".toHaveErrors") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(actual)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message: resultMessage };
}

function toHTMLValidate(
	this: jest.MatcherUtils,
	// @ts-ignore DOM library not available
	actual: string | HTMLElement,
	userConfig?: ConfigData,
	filename?: string
): jest.CustomMatcherResult {
	// @ts-ignore DOM library not available
	if (actual instanceof HTMLElement) {
		actual = actual.outerHTML;
	}

	const defaultConfig = {
		rules: {
			/* jsdom normalizes style so disabling rule when using this matcher or it
			 * gets quite noisy when configured with self-closing */
			"void-style": "off",
		},
	};
	const config = deepmerge(defaultConfig, userConfig || {});

	const actualFilename = filename || this.testPath;
	const htmlvalidate = new HtmlValidate();
	const report = htmlvalidate.validateString(actual, actualFilename, config);
	const pass = report.valid;
	if (pass) {
		return { pass, message: () => "HTML is valid when an error was expected" };
	} else {
		const errors = report.results[0].messages.map(
			(message) => `  ${message.message} [${message.ruleId}]`
		);
		return {
			pass,
			message: () =>
				["Expected HTML to be valid but had the following errors:", ""]
					.concat(errors)
					.join("\n"),
		};
	}
}

function toBeToken(
	this: jest.MatcherUtils,
	actual: any,
	expected: any
): jest.CustomMatcherResult {
	const token = actual.value;

	// istanbul ignore next: TokenMatcher requires "type" property to be set, this is just a failsafe
	if (token.type) {
		token.type = TokenType[token.type];
	}

	// istanbul ignore next: TokenMatcher requires "type" property to be set, this is just a failsafe
	if (expected.type) {
		expected.type = TokenType[expected.type];
	}

	const matcher = expect.objectContaining(expected);
	const pass = this.equals(token, matcher);
	const diffString = diff(matcher, token, { expand: this.expand });
	const message = (): string =>
		this.utils.matcherHint(".toBeToken") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(token)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message };
}

expect.extend({
	toBeValid,
	toBeInvalid,
	toHaveError,
	toHaveErrors,
	toHTMLValidate,
	toBeToken,
});
