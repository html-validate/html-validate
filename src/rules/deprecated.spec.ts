import HtmlValidate from "../htmlvalidate";
import "../matchers";

describe("rule deprecated", () => {
	let htmlvalidate: HtmlValidate;

	beforeAll(() => {
		htmlvalidate = new HtmlValidate({
			root: true,
			rules: { deprecated: ["error", { exclude: ["applet"] }] },
			elements: [
				"html5",
				{
					"custom-deprecated": {
						deprecated: "lorem ipsum",
					},
				},
			],
		});
	});

	it("should not report when regular element is used", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateString("<p></p>");
		expect(report).toBeValid();
	});

	it("should not report when unknown element is used", () => {
		expect.assertions(1);
		const markup = "<missing-meta></missing-meta>";
		const report = htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should not report error for ignored deprecated elements", () => {
		expect.assertions(1);
		const markup = "<applet></applet>";
		const report = htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should report error when deprecated element is used", () => {
		expect.assertions(2);
		const markup = "<marquee>foobar</marquee>";
		const report = htmlvalidate.validateString(markup);
		expect(report).toBeInvalid();
		expect(report).toHaveError("deprecated", "<marquee> is deprecated");
	});

	it("should report error when element with deprecation message is used", () => {
		expect.assertions(2);
		const report = htmlvalidate.validateString("<custom-deprecated>foobar</custom-deprecated>");
		expect(report).toBeInvalid();
		expect(report).toHaveError("deprecated", "<custom-deprecated> is deprecated: lorem ipsum");
	});

	it("smoketest", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateFile("test-files/rules/deprecated.html");
		expect(report.results).toMatchSnapshot();
	});

	describe("metadata variants", () => {
		it.each`
			description              | deprecated                           | message                                      | documentation
			${"boolean"}             | ${true}                              | ${"<my-element> is deprecated"}              | ${"The `<my-element>` element is deprecated and should not be used in new code."}
			${"string"}              | ${"lorem ipsum"}                     | ${"<my-element> is deprecated: lorem ipsum"} | ${"The `<my-element>` element is deprecated and should not be used in new code."}
			${"message"}             | ${{ message: "lorem ipsum" }}        | ${"<my-element> is deprecated: lorem ipsum"} | ${"The `<my-element>` element is deprecated and should not be used in new code."}
			${"documentation"}       | ${{ documentation: "Lorem ipsum." }} | ${"<my-element> is deprecated"}              | ${"The `<my-element>` element is deprecated and should not be used in new code.\n\nLorem ipsum."}
			${"source html5"}        | ${{ source: "html5" }}               | ${"<my-element> is deprecated"}              | ${"The `<my-element>` element is deprecated in HTML 5 and should not be used in new code."}
			${"source html53"}       | ${{ source: "html53" }}              | ${"<my-element> is deprecated"}              | ${"The `<my-element>` element is deprecated in HTML 5.3 and should not be used in new code."}
			${"source whatwg"}       | ${{ source: "whatwg" }}              | ${"<my-element> is deprecated"}              | ${"The `<my-element>` element is deprecated in HTML Living Standard and should not be used in new code."}
			${"source non-standard"} | ${{ source: "non-standard" }}        | ${"<my-element> is deprecated"}              | ${"The `<my-element>` element is deprecated and non-standard and should not be used in new code."}
			${"source lib"}          | ${{ source: "my-lib" }}              | ${"<my-element> is deprecated"}              | ${"The `<my-element>` element is deprecated by my-lib and should not be used in new code."}
		`("$description", ({ deprecated, message, documentation }) => {
			const htmlvalidate = new HtmlValidate({
				rules: { deprecated: "error" },
				elements: [
					"html5",
					{
						"my-element": {
							deprecated,
						},
					},
				],
			});
			const report = htmlvalidate.validateString("<my-element></my-element>");
			expect(report).toHaveError("deprecated", message);
			const context = report.results[0].messages[0].context;
			const doc = htmlvalidate.getRuleDocumentation("deprecated", null, context);
			expect(doc?.description).toEqual(documentation);
		});
	});

	it("should contain documentation", () => {
		expect.assertions(1);
		expect(htmlvalidate.getRuleDocumentation("deprecated")).toMatchSnapshot();
	});

	it("should contain contextual documentation", () => {
		expect.assertions(4);
		expect(
			htmlvalidate.getRuleDocumentation("deprecated", null, {
				tagName: "center",
			})
		).toMatchSnapshot();
		expect(
			htmlvalidate.getRuleDocumentation("deprecated", null, {
				tagName: "blink",
				source: "html5",
			})
		).toMatchSnapshot();
		expect(
			htmlvalidate.getRuleDocumentation("deprecated", null, {
				tagName: "blink",
				source: "html41",
			})
		).toMatchSnapshot();
		expect(
			htmlvalidate.getRuleDocumentation("deprecated", null, {
				tagName: "blink",
				documentation: "extra documentation for <$tagname>",
			})
		).toMatchSnapshot();
	});
});
