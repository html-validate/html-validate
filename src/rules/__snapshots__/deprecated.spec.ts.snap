// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule deprecated smoketest 1`] = `
Array [
  Object {
    "errorCount": 1,
    "filePath": "test-files/rules/deprecated.html",
    "messages": Array [
      Object {
        "column": 1,
        "line": 2,
        "message": "<marquee> is deprecated",
        "offset": 19,
        "ruleId": "deprecated",
        "severity": 2,
        "size": 8,
      },
    ],
    "source": "<p>lorem ipsum</p>
<marquee>foobar</marquee>
",
    "warningCount": 0,
  },
]
`;
